File: word_benchmark/contains.go
package main

import ("flag";"fmt";"io/ioutil";"runtime";"strings";"time")

func contains(needle string, haystack []string, result chan int) int {
	var out []string
	for _, word := range haystack {
		if strings.Contains(word, needle) {
			out = append(out, word)
		}
	}
	result <- len(out)
	return len(out)
}

func getSource(path string) []string {
	file, _ := ioutil.ReadFile(path)
	return strings.Split(string(file), "\n")
}

var word = flag.String("word", "", "Specify the word")
var path = flag.String("path", "", "Specify the path")
var workers = flag.Int("workers", runtime.NumCPU(), "Specify the workers")

func main() {
	flag.Parse()
	numCPUs := runtime.NumCPU()
	runtime.GOMAXPROCS(numCPUs)

	haystack := getSource(*path)
	haystackSize := len(haystack)
	size := haystackSize / *workers
	partitions := make([][]string, haystackSize / size)
	numPartitions := cap(partitions)

	startTime := time.Now()
	startIndex := 0
	endIndex := size
	done := make(chan int, numPartitions)
	for i := 0; i < numPartitions; i++ {
		go contains(*word, haystack[startIndex:endIndex], done)
		startIndex = startIndex + size
		endIndex = startIndex + size
		if endIndex >= haystackSize {
			endIndex = haystackSize - 1
		}
	}

	total := 0
	for i := 0; i < numPartitions; i++ {
		total = total + <-done
	}
	fmt.Printf("Matched: %d\n", total)
	fmt.Printf("Critical section: %f seconds\n", (time.Now().Sub(startTime)).Seconds())
}

File: word_benchmark/contains.py
from optparse import OptionParser
import multiprocessing
import time


def contains(needle, haystack, results):
    out = []
    for word in haystack:
        if needle in word:
            out.append(word)
    results.put(len(out))


def get_haystack(path):
    with open(path) as f:
        return f.read().splitlines()


if __name__ == '__main__':
    parser = OptionParser()
    parser.add_option('-w', '--word', dest='word')
    parser.add_option('-p', '--path', dest='path')
    parser.add_option('-k', '--workers', dest='workers', type="int")
    (options, args) = parser.parse_args()

    num_cpus = multiprocessing.cpu_count()
    if not options.workers:
        options.workers = num_cpus
    haystack = get_haystack(options.path)
    size = len(haystack) / options.workers

    start = time.time()
    partitions = zip(*[iter(haystack)] * size)

    results = multiprocessing.Queue()
    jobs = []
    for partition in partitions:
        p = multiprocessing.Process(target=contains, args=(options.word, partition, results))
        jobs.append(p)
        p.start()

    total = 0
    for job in jobs:
        job.join()
        total += results.get()
    print "Matched: {0}".format(total)
    print "Critical section: {0:f} seconds".format(time.time() - start)
